### PR TITLE
(CDAP-727) Added support of HttpContentProducer to Service

### PR DESCRIPTION
**Pull request by chtyim**
- Addition to public API
  - HttpContentProducer for user to implement
  - Multiple send methods in HttpServiceResponder for sending with Location or HttpContentProducer
- Added support of HttpContentProducer in the service implementation
  - Pipe calls from netty-http BodyProducer to HttpContentProducer
  - Unify handling of service context capture and release
  - Unify handling of dismissing transaction context
- Modified examples to use the new send method in HttpServiceResponder to send file back to client

### CODE REVIEW

**Comment by chtyim** [comment]
Build passed: http://builds.cask.co/browse/CDAP-RBT573-4

File: cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+package co.cask.cdap.api.service.http;
+
+import co.cask.cdap.api.Transactional;
+import co.cask.cdap.api.TxRunnable;
+import co.cask.cdap.api.dataset.Dataset;
+
+import java.nio.ByteBuffer;
+
+/**
+ * Instance of this class is for producing HTTP response body in small chunks to avoid running out of memory
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java]: This class is for producing an HTTP response ..

File: cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ * when responding with large body. An instance of this class can be provided to one of the {@code send}
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java]: with a large body.

File: cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+  /**
+   * Returns the length of the content or {@code -1L} if the length is unknown.
+   */
+  public long getContentLength() {
+    return -1L;
+  }
+
+  /**
+   * This method will get invoked when a new chunk of data is going to send to the client.
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java]: This method provides a new chunk chunk of data to be sent to the client.

File: cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+   * @param transactional for executing a {@link TxRunnable} in a single transaction.
+   * @return a {@link ByteBuffer} containing the next chunk of bytes to be sent to the client.
+   *         If the returned {@link ByteBuffer} is empty ({@link ByteBuffer#hasRemaining()} == {@code false}),
+   *         it signals that's the end of the response body.
+   * @throws Exception if there is any error
+   */
+  public abstract ByteBuffer nextChunk(Transactional transactional) throws Exception;
+
+  /**
+   * This method will get invoked after the last chunk of data is sent. This method is always
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java]: It is always

File: cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+  public abstract void onFinish() throws Exception;
+
+  /**
+   * This method will get invoked when there is an error while sending body chunks.
+   *
+   * Any issues related to network as well as any {@link Exception Exceptions} raised
+   * from either {@link #nextChunk(Transactional)}
+   * or {@link #onFinish()} methods will have this method invoked.
+   *
+   * This method is always executed inside a single transaction.
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java]: what happens if this method itself throws an exception?

File: cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpServiceResponder.java [hunk 6]
```diff
@@ -139,11 +142,76 @@
...
   void send(int status, ByteBuffer content, String contentType, Iterable<? extends Map.Entry<String, String>> headers);
+
+  /**
+   * Sends response back to client with response body produced by the given {@link HttpContentProducer}.
+   *
+   * @param status status of the HTTP response
+   * @param producer a {@link HttpContentProducer} to produce content to be sent back
+   * @param contentType type of content
+   */
+  void send(int status, HttpContentProducer producer, String contentType);
```
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpServiceResponder.java]: why is it that the content length is given by the producer itself, whereas the content type is a separate parameter?

Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpServiceResponder.java]: This helps one easier to implement reusable `HttpContentProducer`. E.g. he can implements a `FileHttpContentProducer` that can read from any file to produce the content, but only the caller knows what file it is and hence be able to provide the correct content type.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java [hunk 7]
```diff
@@ -95,14 +96,35 @@ protected final HttpServiceRequest wrapRequest(HttpRequest request) {
...
   @SuppressWarnings("unused")
-  protected final DelayedHttpServiceResponder wrapResponder(HttpResponder responder) {
+  protected final DelayedHttpServiceResponder wrapResponder(HttpResponder responder,
+                                                            final TransactionContext txContext) {
+
+
     MetricsContext collector = this.metricsContext;
     HttpServiceContext serviceContext = context.getServiceContext();
-    if (serviceContext != null && serviceContext.getSpecification() != null) {
+    Preconditions.checkState(serviceContext instanceof TransactionalHttpServiceContext,
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java]: can that ever happen?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java [hunk 8]
```diff
@@ -95,14 +96,35 @@ protected final HttpServiceRequest wrapRequest(HttpRequest request) {
...
       collector = metricsContext.childContext(Constants.Metrics.Tag.HANDLER,
                                               serviceContext.getSpecification().getName());
     }
-    return new DelayedHttpServiceResponder(responder, collector);
+
+    return new DelayedHttpServiceResponder(responder, new BodyProducerFactory() {
+      @Override
+      public BodyProducer create(HttpContentProducer contentProducer, TransactionalHttpServiceContext serviceContext) {
+        final ClassLoader programContextClassLoader = new CombineClassLoader(
+          null, ImmutableList.of(contentProducer.getClass().getClassLoader(), getClass().getClassLoader()));
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java]: why is this needed?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java [hunk 9]
```diff
@@ -95,14 +96,35 @@ protected final HttpServiceRequest wrapRequest(HttpRequest request) {
...
+
+        Transactional transactional = createTransactional(txContext, programContextClassLoader, serviceContext);
+
+        // Capture the context since we need to keep it till the end of the content producing.
+        // This method only gets call when HttpContentConsumer is not used
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java]: this comment confuses me

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java (hunk 7 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java]: Not with the current implementation. It was due to some old code logic so that `DelegatorContext.getServiceContext` returns a `HttpServiceContext` instead of `TransactionalHttpServiceContext`, hence with the check to guard against potential bug. Will do a refactoring in a separate PR to change `DelegatorContext` to always return `TransactionalHttpServiceContext`.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java [hunk 10]
```diff
@@ -118,11 +140,60 @@ protected final DelayedHttpServiceResponder wrapResponder(HttpResponder responde
   protected final BodyConsumer wrapContentConsumer(HttpContentConsumer consumer,
                                                    DelayedHttpServiceResponder responder,
                                                    TransactionContext txContext) {
+    Preconditions.checkState(!responder.hasBufferedResponse(),
+                             "Response cannot be sent when using HttpContentConsumer.");
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java]: is this a misleading error message?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java (hunk 8 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java]: It is needed for this bug.. https://issues.cask.co/browse/CDAP-2972

We may not need this if the dataset framework is setting the classloader correctly in all scenario (see https://issues.cask.co/browse/CDAP-4189 and related JIRA).

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java (hunk 9 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java]: When `HttpContentConsumer` is used, a different `BodyProducerFactory` will be used, hence this code path will not be executed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java (hunk 10 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java]: It tries to capture this misuse.

```java
HttpContentConsumer handle(HttpServiceRequest request, HttpServiceResponder responder) {
  responder.sendStatus(200);
  return new HttpContentConsumer() {
     ...
  };
}
```

because response should be sent from the `onFinish` method in the `HttpContentConsumer`. Any recommendation on what the error message should be?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java (hunk 8 above)
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java]: oh the content producer can be a cdap class, like the one for Location...

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java [hunk 11]
```diff
@@ -133,9 +127,15 @@ public void setTransactionFailureResponse(Throwable t) {
...
       @Override
-      public void execute() {
+      public void execute(boolean keepAlive) {
         // no-op
       }
+
+      @Override
+      public boolean hasContentProducer() {
+        // Always release the context at the end since it's not possible to send with a content producer
+        return false;
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java]: should you then validate that in doSend() that the contentProducer is null?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java [hunk 12]
```diff
@@ -160,51 +160,40 @@ public void run(DatasetContext context) throws Exception {
...
-      } finally {
-        ClassLoaders.setContextClassLoader(oldClassLoader);
-      }
-    } catch (Throwable t) {
-      txContext.abort(new TransactionFailureException("Exception raised from TxRunnable.run()", t));
+  private static final class ErrorBodyProducerFactory implements BodyProducerFactory {
+
+    @Override
+    public BodyProducer create(HttpContentProducer contentProducer, TransactionalHttpServiceContext serviceContext) {
+      // It doesn't what it returns as it'll never get used
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java]: it doesn't matter?

Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java]: If it does not get used, could you just return null?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java (hunk 11 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java]: For this case, when the `handleError` method gets called from netty-http, a response has already sent (by the library), so it doesn't really matter what the `doSend` method get (and in fact, send->doSend only gets call from the `onError` method, which is a no-op anyway).

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java (hunk 12 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java]: I can in the current implementation. However, I didn't because the interface contract doesn't allow returning `null`. If any change in the implementation of the `DelayedHttpServiceResponder` that have this method called, then it's not good to have `null` returned.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java [hunk 13]
```diff
@@ -158,27 +161,72 @@ public FileUploadHandler(File outputDir) {
...
-          responder.sendStatus(200);
-        }
+    @Path("/download/{file}")
+    @GET
+    public void download(HttpServiceRequest request,
+                         HttpServiceResponder responder,
+                         @PathParam("file") String file) {
+      Location location = Locations.toLocation(new File(outputDir, file));
+      try {
+        responder.send(200, location, "text/plain");
```
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java]: this makes me think. It is so easy to send the content of a location. Can we make it equally easy to stream a request body to a location?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java [hunk 14]
```diff
@@ -158,27 +161,72 @@ public FileUploadHandler(File outputDir) {
...
+    @POST
+    public HttpContentConsumer uploadDownload(HttpServiceRequest request,
+                                              HttpServiceResponder responder,
+                                              @PathParam("file") String file) throws IOException {
+      final File targetFile = new File(outputDir, file);
+      return new FileContentConsumer(targetFile) {
         @Override
-        public void onError(HttpServiceResponder responder, Throwable failureCause) {
-          LOG.error("Failed when handling upload", failureCause);
+        protected void response(HttpServiceResponder responder) throws IOException {
```
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java]: this first reads the entire request body, then sends back the response. Is it actually possible to interleave the receiving and sending? For example, you read a request that has a huge list of keys in its body; for each key you look it up in some table and send key:value back. In that case, you would want to start sending at the same as you are still receiving

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java [hunk 15]
```diff
@@ -261,33 +309,119 @@ protected FileUploadHandler createHandler() {
...
         String.format("http://%s:%d/content/upload/test.txt",
                       bindAddress.getHostName(), bindAddress.getPort())).openConnection();
+      try {
+        urlConn.setReadTimeout(2000);
+        urlConn.setDoOutput(true);
+        urlConn.setRequestMethod("PUT");
+        // Set to use default chunk size
+        urlConn.setChunkedStreamingMode(-1);
+
+        // Write over 150MB of data
```
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java]: does it have to be that large? this will write 400MB of data to temp files... Isn't 20MB enough to test chunking?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java (hunk 13 above)
Comment (chtyim) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java]: We can provide something like `PartitionedFIleSetContentConsumer` in cdap-api to do so.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java [hunk 16]
```diff
@@ -261,33 +309,119 @@ protected FileUploadHandler createHandler() {
...
+    final File outputDir = TEMP_FOLDER.newFolder();
+    HttpHandler httpHandler = factory.createHttpHandler(
+      TypeToken.of(FileHandler.class), new AbstractDelegatorContext<FileHandler>() {
+        @Override
+        protected FileHandler createHandler() {
+          return new FileHandler(outputDir);
         }
+      });
+
+    NettyHttpService service = NettyHttpService.builder()
```
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java]: so you are testing this directly against a netty-http. Why not against a CDAP service?

Comment (chtyim) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java]: This test class is mainly for testing the generated class correctness. It's much easier to debug if something goes wrong in the bytecode generation.

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifeCycleTestRun.java [hunk 17]
```diff
@@ -256,6 +275,142 @@ public Long call() throws Exception {
...
+
+      // Wait for completion
+      Futures.successfulAsList(completions).get(10, TimeUnit.SECONDS);
+
+      // Verify the upload result
+      for (ListenableFuture<Integer> future : completions) {
+        Assert.assertEquals(200, future.get().intValue());
+      }
+
+      // Get the states again, it should still be 6 same instances
```
Comment (anew) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifeCycleTestRun.java]: can they expire before you make this call, on a slow machine? That would lead to a flaky test...

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java (hunk 15 above)
Comment (chtyim) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java]: This is due to the default 150MB memory buffer used by netty-http server. We wrote > 150M of data to make sure we go over that 150MB limit without failure.

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifeCycleTestRun.java (hunk 17 above)
Comment (chtyim) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifeCycleTestRun.java]: The default expiration is 60 seconds, which seems unlikely even on slow machine.

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifecycleApp.java [hunk 18]
```diff
@@ -79,7 +101,62 @@ public void onFinish(HttpServiceResponder responder) throws Exception {
...
+      final HttpServiceResponder finalResponder = responder;
+      return new HttpContentConsumer() {
+        @Override
+        public void onReceived(ByteBuffer chunk, Transactional transactional) throws Exception {
+          // no-op
+        }
+
+        @Override
+        public void onFinish(HttpServiceResponder responder) throws Exception {
+          // Use the closure responder instead of the given responder, it should fail
```
Comment (anew) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifecycleApp.java]: this is tricky. We need to make sure the response it very meaningful when this happens. Also make the docs very clear that you can only use the responder provided with the callback, not the original one.

Comment (chtyim) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifecycleApp.java]: It is documented in the `HttpContentConsumer` https://github.com/caskdata/cdap/blob/release/3.3/cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentConsumer.java#L86

The error message is part of this PR https://github.com/caskdata/cdap/pull/4851/files#diff-5247f6df941a76487fca97cb33257a17R87

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifecycleApp.java [hunk 19]
```diff
@@ -79,7 +101,62 @@ public void onFinish(HttpServiceResponder responder) throws Exception {
...
+          finalResponder.sendStatus(200);
```
Comment (anew) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifecycleApp.java]: So this will throw an exception, which gets caught and the onError() is invoked. What if onError also uses the original responder? That will throw another exception, which will be logged and the connection gets closed without response?

File: cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpServiceResponder.java (hunk 6 above)
Comment (anew) [inline cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpServiceResponder.java]: ok, that makes sense.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java (hunk 7 above)
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java]: ok

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifecycleApp.java (hunk 19 above)
Comment (chtyim) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifecycleApp.java]: It will be logged with a 500 respond error message. The message of the original exception should be part of the response message.

File: cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java (hunk 5 above)
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java]: It will only be logged and connection get closed. There nothing much can be done.

Actually notice a bug in the current implementation. The exception should be propagated to back to the netty-http library so that the connection get closed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java (hunk 10 above)
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java]: Perhaps "HttpContentConsumer may not be used after a response has already been sent"

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java (hunk 11 above)
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java]: ok. intricate :)

File: cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java (hunk 1 above)
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java]: fixed

File: cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java (hunk 2 above)
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java]: fixed.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java (hunk 12 above)
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/BodyConsumerAdapter.java]: +1 for complying with the contract

File: cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java (hunk 3 above)
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java]: fixed.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java (hunk 13 above)
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java]: Or just LocationContentConsumer?

File: cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+  }
+
+  /**
+   * This method will get invoked when a new chunk of data is going to send to the client.
+   * It is guaranteed that no concurrent calls to this method will be made.
+   * The implementation can reuse the same {@link ByteBuffer} instance
+   * across multiple calls to this method.
+   *
+   * <p>
+   * Access to transactional {@link Dataset Datasets} should be done through the
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java]: fixed.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java (hunk 15 above)
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java]: or could we tune that buffer to a smaller size for unit tests?

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java (hunk 16 above)
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java]: ok, makes sense.

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifeCycleTestRun.java (hunk 17 above)
Comment (anew) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifeCycleTestRun.java]: I hope so.

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifecycleApp.java (hunk 18 above)
Comment (anew) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifecycleApp.java]: oh cool that is excellent.

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifecycleApp.java (hunk 19 above)
Comment (anew) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifecycleApp.java]: Hmm but the code here: https://github.com/caskdata/cdap/pull/4851/files#diff-c1f37562036eae5e370dfab1d3cd720dR100 assumes that the response has already been sent. So there will be no response. Or what am I missing?

Comment (chtyim) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifecycleApp.java]: There is a bug in the current implementation. See my comment in https://github.com/caskdata/cdap/pull/4851/files#r49655451

File: cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java (hunk 4 above)
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/service/http/HttpContentProducer.java]: Fixed

File: cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifecycleApp.java (hunk 19 above)
Comment (anew) [inline cdap-unit-test/src/test/java/co/cask/cdap/test/app/ServiceLifecycleApp.java]: oh I missed that. Should have refreshed my page more.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java (hunk 10 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/service/http/AbstractHttpHandlerDelegator.java]: Fixed.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java (hunk 14 above)
Comment (chtyim) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java]: It common that http client doesn't start reading the response before completely sending out the request. If we start sending the respond while receiving a large request body, we might end up filling up the network buffer and causing deadlock.

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java (hunk 15 above)
Comment (chtyim) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/runtime/service/http/HttpHandlerGeneratorTest.java]: Yes. Fixed.

**Comment by chtyim** [comment]
Addressed all comments. Please take a look again.

**Comment by anew** [comment]
LGTM. Excellent work :gem:

**Comment by anew** [comment]
LGTM

**Comment by chtyim** [comment]
Build passed: http://builds.cask.co/browse/CDAP-RBT573-8

**Comment by chtyim** [comment]
Squashing commits and merging.
